### ISSUE TITLE
LLVM cannot constant fold memchr with local buffer

### ISSUE THREAD

**Issue by philnik777** [llvm/llvm-project#114668]
LLVM is currently unable to constant fold calls `memchr` when a local buffer is passed. I noticed this while looking into libc++'s `std::find` optimizations.

This is a simple example where Clang doesn't constant fold the function even though it should be able to do that trivially: ([Godbolt](https://godbolt.org/z/xs7sc4TYz))
```c++
auto test() {
  char buffer[] = "Banane";
  return __builtin_memchr(buffer, 'a', __builtin_strlen(buffer)) == nullptr;
}
```

I think the problem is that `__builtin_memchr` can escape the pointer, so the buffer isn't recognized as immutable, resulting in the constant folder failing.

I'm in no way an expert in compiler optimizations, but I think it would be possible to check whether the pointer actually escapes, and if not, mark the pointer as `noescape` to at least partially solve this problem.

**Comment by nikic** [comment]
Yes, missing nocapture is the issue. It folds if we manually add it to the IR: https://llvm.godbolt.org/z/Ed553Yvxd

It should be easy to infer nocapture in SimplifyLibCalls for memchr calls that are only compared against nullptr.

The case where the return value is used for more than that would be harder though.

Tracking return-only captures (part of https://discourse.llvm.org/t/rfc-improvements-to-capture-tracking/81420) would probably also handle your case as a side effect.

**Comment by sidnarsipur** [comment]
Hey, can I be assigned this?

**Comment by NeyagapulaSiddhartha** [comment]
hey if no one is working on this , I would like to be assigned !!

**Comment by nikic** [comment]
I haven't tested it, but I'd expect that https://github.com/llvm/llvm-project/pull/132007 fixes this. It's currently blocked on https://github.com/llvm/llvm-project/issues/131168 though.
